### ISSUE TITLE
Does not fetch the best quality link

### ISSUE THREAD

**Issue by WadeShuler** [Athlon1600/youtube-downloader#137]
This doesn't fetch the best quality video source, it simply grabs the first one from the combined (video/audio) list returned.

You need to sort through the combined results, compare the widths, and return the one with the largest size.

Something as simple as this in DownloadOptions:

```php
public function getBestCombinedFormat()
{
    $combined = $this->getCombinedFormats();

    $bestResult = null;
    $bestWidth = 0;

    foreach($combined as $test) {
        if ( isset($test->width) && ($test->width > $bestWidth) ) {
            $bestWidth = $test->width;
            $bestResult = $test;
        }
    }

    return $bestResult;
}
```

Then update in the try/catch in video_info.php:

```php
$links = $youtube->getDownloadLinks($url);
$best = $links->getBestCombinedFormat();
```

**Comment by 89z** [comment]
YouTube-DL does this, and I think its a terrible idea. You dont want to just sort the results, you want to also filter them. For example, with this 10 minute video [1], here are some of the results:

~~~
itag 315, height 2160, 26.8 mb/s, 1.5 GB, video/webm; codecs="vp9"
itag 401, height 2160, 24.2 mb/s, 860.1 MB, video/mp4; codecs="av01.0.13M.08"
itag 308, height 1440, 13.3 mb/s, 545.1 MB, video/webm; codecs="vp9"
itag 400, height 1440, 16.0 mb/s, 486.1 MB, video/mp4; codecs="av01.0.12M.08"
itag 299, height 1080, 6.7 mb/s, 271.2 MB, video/mp4; codecs="avc1.64002a"
itag 303, height 1080, 4.5 mb/s, 204.0 MB, video/webm; codecs="vp9"
itag 399, height 1080, 4.9 mb/s, 164.6 MB, video/mp4; codecs="av01.0.09M.08"
itag 136, height 720, 2.6 mb/s, 94.2 MB, video/mp4; codecs="avc1.4d401f"
itag 247, height 720, 1.5 mb/s, 75.9 MB, video/webm; codecs="vp9"
itag 298, height 720, 4.1 mb/s, 159.4 MB, video/mp4; codecs="avc1.4d4020"
itag 302, height 720, 2.7 mb/s, 124.3 MB, video/webm; codecs="vp9"
itag 398, height 720, 3.3 mb/s, 98.2 MB, video/mp4; codecs="av01.0.08M.08"
~~~

I think in general, most people would be fine with defaulting to 720p, which in this case is 125 MB. If we just blindly take the highest quality, youre downloading 1.5 GB instead. Some people may need or want the 4K result, but I dont think it should be the default.

1. https://www.youtube.com/watch?v=aqz-KE-bpKQ

**Comment by WadeShuler** [comment]
I understand what you are saying, but I would imagine most people want the highest quality available in most situations.

If I lost my backup HD and I need to download ALL of my old YouTube videos, why would I want an even worse quality version? When you upload it the first time, YouTube compresses it. So it's already a quality hit compared to your original. In the example above, the 1.5GB is as close to your original quality as you can get. So in the sense of using a tool like this to recover copies of all of your old uploaded videos (hard drive failed or whatever), you would always want the best quality version you can get to be as close to the original.

Downloading 1.5GB isn't a problem for most people's internet today.

User configurable options would work. Maybe allow them to take the best quality/largest size available regardless, or choose an ideal size (like 720 or 1080), or a file size limit (ie: don't take videos 1GB or larger). Instead of starting the download, you could return the results and let them choose which one they want.

It is always best to provide the developer with configurable options, or the user with choices, instead of forcing your choice on them. 1.5GB may seem large to you, but isn't anything for others to download.

**Comment by 89z** [comment]
> I understand what you are saying, but I would imagine most people want the highest quality available in most situations.

Nope. Most people probably just want a quick download of something thats "good enough". Sure, some people are going to want the absolute best quality, but in general the size tradeoff is not worth the slight quality increase.

> If I lost my backup HD and I need to download ALL of my old YouTube videos, why would I want an even worse quality version?

OK, but thats **your** perspective. You have to think of the common usecase, which is just get a quick pretty good quality download. Think about it like this: regarding food, you can literally make your own bread if you want to [1], but probably the average person is just going to want to buy their bread, because its quicker.

> So in the sense of using a tool like this to recover copies of all of your old uploaded videos (hard drive failed or whatever), you would always want the best quality version you can get to be as close to the original.

Again, thats how **you** want to use it. Nothing is stopping you from getting the best quality from YouTube-DL or this tool, but it shouldnt be the default option.

> Downloading 1.5GB isn't a problem for most people's internet today.

Now youre just rationalizing. Just because internet is faster than it used to be, doesnt justify a 10 fold increase in file size. You have to compare the videos on their own, and weight the cost/benefit of each option. You know that the larger files lose on their own merit, so youre reaching for any argument now, even a bad one.

> User configurable options would work. Maybe allow them to take the best quality/largest size available regardless, or choose an ideal size (like 720 or 1080), or a file size limit (ie: don't take videos 1GB or larger). Instead of starting the download, you could return the results and let them choose which one they want.

Or we can just do what YouTube does. Default to 720p, then if someone want better quality, all they have to do is change the setting on the video itself, or I think you can change the default in user options as well. But in the absence of any user configuration, 4K **is not** a sensible default.

> It is always best to provide the developer with configurable options, or the user with choices, instead of forcing your choice on them. 1.5GB may seem large to you, but isn't anything for others to download.

No one is doing that. Every single tool ive seen, from YouTube-DL, to this tool, to my own tool [2], all have a default, and all have an option to change the quality. YouTube-DL defaulting to 4K was a dumb decision, one that YouTube itself doesnt do, and no other tool should either.

1. https://www.youtube.com/watch?v=Jizr6LR83Kk
2. https://github.com/89z/mech

**Comment by WadeShuler** [comment]
Wow, extremely rude and hypocritical.

I obviously created a NEW method to download the best quality, which would allow the developer to use if they CHOOSE to.

NOWHERE did I say to make it get the best quality by default.

> all have a default, and all have an option to change the quality

Yours does not seem to allow anyone to choose anything other than the quality Hitler decided they can have. Your tool grabs the results from YouTube, filters the combined (video and audio), then returns the first one without taking any dimensions or size into account. You get whatever is first in the array.

> Nope. Most people probably just want a quick download of something thats "good enough". Sure, some people are going to want the absolute best quality, but in general the size tradeoff is not worth the slight quality increase.

> OK, but thats your perspective. You have to think of the common usecase, which is just get a quick pretty good quality download. 

> Again, thats how you want to use it.

> Now youre just rationalizing. Just because internet is faster than it used to be, doesnt justify a 10 fold increase in file size. You have to compare the videos on their own, and weigh the cost/benefit of each option. You know that the larger files lose on their own merit, so youre reaching for any argument now, even a bad one.

You did exactly what you accused me of, rationalizing based on your own opinion. Therefore nobody else has the option because of your opinionated "trade-off". What you said, just makes you look like a hypocrite.

Most of us with actual high speed internet, have no concern about file size. We download them in seconds, or maybe a few minutes at most.

You created a script that downloads videos from YouTube for other developers to use on their own projects. However, you have made the ultimate decision in what is best for them based on your own opinion. You made it to be something that if anyone wants anything different than what you decide, they have to customize it themselves. That is just bad design. It seriously isn't too difficult to add some config options, or some methods, to allow flexibility.

**Comment by ewwink** [comment]
sorter code no need loop, it already sorted lowest to highest

    public function getBestCombinedFormat()
    {
        $combined = $this->getCombinedFormats();
        $count = count($combined);
        echo $count ? $combined[$count-1] : null;
    }
